#**Text Preprocessing in Natural Language Processing (NLP)**

<div style="font-family: 'Gill Sans', 'Open Sans', sans-serif;">

## **1. Lowercasing**

### **What is this technique?**
Lowercasing is the process of converting all characters in a text to lowercase.

**Example:**
"I HAVE a meeting" → "i have a meeting"

### **Why is it used?**
- Reduces vocabulary size (e.g., "Hello" and "hello" become the same)
- Improves model efficiency
- Avoids duplicate word representations

</div>

In [3]:
raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!",
    "Hi!!! How are you doing today??? 😊",
    "Complete the project ASAP!!! Deadline is 25/12/2025.",
    "Let's catch up at Starbucks, Chennai.",
    "Call mom @ 7pm sharp.",
    "Meeting rescheduled from 5pm to 6:30pm.",
    "Send the EMAIL to HR_department immediately.",
    "Book flight tickets from Chennai to Delhi.",
    "Wake me up at 6 in the morning.",
    "Don't forget to bring the documents!!!",
    "Can we meet next Friday at 4pm?"
]

def to_lowercase(text):
    return text.lower()

lowercased_texts = [to_lowercase(text) for text in raw_texts]

In [5]:
for before, after in zip(raw_texts, lowercased_texts):
    print("BEFORE :", before)
    print("AFTER  :", after)
    print("-" * 50)

BEFORE : I HAVE a meeting @ 3pm!!!
AFTER  : i have a meeting @ 3pm!!!
--------------------------------------------------
BEFORE : Hey John, can u pls call me tomorrow??
AFTER  : hey john, can u pls call me tomorrow??
--------------------------------------------------
BEFORE : Reminder: Pay electricity bill before 10th Sept.
AFTER  : reminder: pay electricity bill before 10th sept.
--------------------------------------------------
BEFORE : Schedule meeting with Dr. Smith at 10:30 AM.
AFTER  : schedule meeting with dr. smith at 10:30 am.
--------------------------------------------------
BEFORE : Buy 2kg apples, 3 bananas & some milk!!
AFTER  : buy 2kg apples, 3 bananas & some milk!!
--------------------------------------------------
BEFORE : Hi!!! How are you doing today??? 😊
AFTER  : hi!!! how are you doing today??? 😊
--------------------------------------------------
BEFORE : Complete the project ASAP!!! Deadline is 25/12/2025.
AFTER  : complete the project asap!!! deadline is 25/12/

<div style="font-family: 'Gill Sans', 'Segoe UI', sans-serif;">

### **What changed?**
- All capital letters are converted into small letters.
- Example:
  "EMAIL" → "email"  
  "ASAP" → "asap"

### **Did the meaning change?**
- No, the meaning stays the same.
- Only the writing style (capital → small letters) is changed.

### **Is this good or bad for NLP?**

Good:
- Reduces duplicate words (like "Hello" and "hello")
- Makes text more consistent and easier for models to understand

Small problem:
- Sometimes we lose special meaning  
  (Example: "ASAP" shows urgency, but "asap" looks normal)



</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', sans-serif;">

## **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No, in most cases we should not remove them.

- Numbers like "3pm" give important information about time.
- If we remove numbers, the sentence may lose its actual meaning.
- For example, in reminders or schedules, time is very important.

### **2. Should we remove names like "John"?**
No, we should not remove names.

- Names are very important in text data.
- They are used in tasks like Named Entity Recognition (NER).
- Removing names can reduce the performance of applications like chatbots.

### **3. Which preprocessing step can cause problems in entity detection?**
Lowercasing can create issues.

- Example:
  "John" → "john"

- When we convert to lowercase, we lose capitalization.
- Capital letters help models identify names, places, and other important entities.
- So, this can make it harder for the model to recognize them correctly.


##--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', sans-serif;">

# **2. Removing Punctuation**

### **What is this technique?**
Removing punctuation means removing special characters from text.

- These include symbols like: ., !, ?, @, #, &, etc.
- These symbols usually do not add much meaning in many NLP tasks.

**Example:**
"Hello!!! How are you?" → "Hello How are you"

### **Why is it used?**

- It helps remove unwanted noise from the text.
- Makes the sentence cleaner and easier to process.
- Helps models focus only on important words.
- Improves consistency in the dataset.


</div>

In [7]:
import string

raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!",
    "Hi!!! How are you doing today??? 😊",
    "Complete the project ASAP!!! Deadline is 25/12/2025.",
    "Let's catch up at Starbucks, Chennai.",
    "Call mom @ 7pm sharp.",
    "Meeting rescheduled from 5pm to 6:30pm.",
    "Send the EMAIL to HR_department immediately.",
    "Book flight tickets from Chennai to Delhi.",
    "Wake me up at 6 in the morning.",
    "Don't forget to bring the documents!!!",
    "Can we meet next Friday at 4pm?"
]

def remove_punctuation(text):
    return text.translate(str.maketrans('', '', string.punctuation))

no_punct_texts = [remove_punctuation(text) for text in raw_texts]

In [8]:
for before, after in zip(raw_texts, no_punct_texts):
    print("BEFORE :", before)
    print("AFTER  :", after)
    print("-" * 50)

BEFORE : I HAVE a meeting @ 3pm!!!
AFTER  : I HAVE a meeting  3pm
--------------------------------------------------
BEFORE : Hey John, can u pls call me tomorrow??
AFTER  : Hey John can u pls call me tomorrow
--------------------------------------------------
BEFORE : Reminder: Pay electricity bill before 10th Sept.
AFTER  : Reminder Pay electricity bill before 10th Sept
--------------------------------------------------
BEFORE : Schedule meeting with Dr. Smith at 10:30 AM.
AFTER  : Schedule meeting with Dr Smith at 1030 AM
--------------------------------------------------
BEFORE : Buy 2kg apples, 3 bananas & some milk!!
AFTER  : Buy 2kg apples 3 bananas  some milk
--------------------------------------------------
BEFORE : Hi!!! How are you doing today??? 😊
AFTER  : Hi How are you doing today 😊
--------------------------------------------------
BEFORE : Complete the project ASAP!!! Deadline is 25/12/2025.
AFTER  : Complete the project ASAP Deadline is 25122025
----------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', sans-serif;">

### **What changed?**
- All punctuation symbols are removed from the text.

**Examples:**
- "meeting @ 3pm!!!" → "meeting  3pm"  
- "don't" → "Dont"  
- "let's" → "Lets"  

### **Did the meaning change?**
- Yes, a little bit.

- Some words lose their correct form  
  (Example: "don't" becomes "Dont")  
- Time formats can also change  
  (Example: "10:30" becomes "1030")  

### **Is this good or bad for NLP?**

Good:
- Removes unnecessary symbols (noise)
- Makes text simpler and cleaner

Bad:
- Important symbols may be lost
- Contractions get broken (don't → dont)
- Structure of data can change (like time, abbreviations)


</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', sans-serif;">

## **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No, we should not remove them.

- Numbers give important information like time and quantity.
- Removing them can change the meaning of the sentence.

### **2. Should we remove names like "John"?**
No, we should not remove names.

- Names are important in NLP tasks.
- They are used to identify people, places, and organizations.
- Removing them can reduce model accuracy.

### **3. Which preprocessing step can affect entity extraction?**
Removing punctuation can cause problems.

**Examples:**
- "Dr. Smith" → "Dr Smith"  
- "10:30 AM" → "1030 AM"  
- "HR_department" → "HRdepartment"  

### **What problems can happen?**
- Difficult to recognize names correctly  
- Time format gets disturbed  
- Organization names may not be detected properly  


</div>

#----------------------------------------------------------------------------------------------------------------------------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', sans-serif;">

# **3. Tokenization**

### **What is this technique?**
Tokenization means breaking text into smaller parts called tokens.  

- Tokens can be words, sentences, or even individual characters.  
- This makes the text easier for computers to process.

**Example:**  
"I love NLP" → ["I", "love", "NLP"]

### **Why is it used?**
- Converts text into a clear, structured format  
- Helps models understand each word separately  
- It is usually the first step in most NLP pipelines

### **Final point:**
Tokenization is important because models cannot understand raw text directly.  

</div>

In [9]:
raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!",
    "Hi!!! How are you doing today??? 😊",
    "Complete the project ASAP!!! Deadline is 25/12/2025.",
    "Let's catch up at Starbucks, Chennai.",
    "Call mom @ 7pm sharp.",
    "Meeting rescheduled from 5pm to 6:30pm.",
    "Send the EMAIL to HR_department immediately.",
    "Book flight tickets from Chennai to Delhi.",
    "Wake me up at 6 in the morning.",
    "Don't forget to bring the documents!!!",
    "Can we meet next Friday at 4pm?"
]

def tokenize_text(text):
    return text.split()

tokenized_texts = [tokenize_text(text) for text in raw_texts]

In [10]:
for before, after in zip(raw_texts, tokenized_texts):
    print("BEFORE :", before)
    print("AFTER  :", after)
    print("-" * 50)

BEFORE : I HAVE a meeting @ 3pm!!!
AFTER  : ['I', 'HAVE', 'a', 'meeting', '@', '3pm!!!']
--------------------------------------------------
BEFORE : Hey John, can u pls call me tomorrow??
AFTER  : ['Hey', 'John,', 'can', 'u', 'pls', 'call', 'me', 'tomorrow??']
--------------------------------------------------
BEFORE : Reminder: Pay electricity bill before 10th Sept.
AFTER  : ['Reminder:', 'Pay', 'electricity', 'bill', 'before', '10th', 'Sept.']
--------------------------------------------------
BEFORE : Schedule meeting with Dr. Smith at 10:30 AM.
AFTER  : ['Schedule', 'meeting', 'with', 'Dr.', 'Smith', 'at', '10:30', 'AM.']
--------------------------------------------------
BEFORE : Buy 2kg apples, 3 bananas & some milk!!
AFTER  : ['Buy', '2kg', 'apples,', '3', 'bananas', '&', 'some', 'milk!!']
--------------------------------------------------
BEFORE : Hi!!! How are you doing today??? 😊
AFTER  : ['Hi!!!', 'How', 'are', 'you', 'doing', 'today???', '😊']
-------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', sans-serif;">

### **What changed?**
- The sentence is split into individual words, called tokens.  

**Example:**  
"I HAVE a meeting @ 3pm!!!" → ["I", "HAVE", "a", "meeting", "@", "3pm!!!"]

### **Did the meaning change?**
- No, the meaning stays the same.  
- Only the format changed from a full sentence to a list of words.

### **Is this good or bad for NLP?**

Good:
- It is an essential step for NLP  
- Helps models process text word by word  
- Converts text into units that are easier to analyze

Limitation:
- Punctuation is still attached (e.g., "3pm!!!")  
- Contractions remain as they are (e.g., "Don't")  

Tip:  
Better tokenization methods like NLTK or spaCy handle punctuation and contractions more accurately.

</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

### **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No  
- Values like "3pm", "7pm", "10:30", and "25/12/2025" are important  
- They give information about time and date  
- If we remove them, meaning will be lost  

### **2. Should we remove names like "John"?**
No  
- Names like "John", "Smith", "Chennai", and "Delhi" are important  
- These are called entities (person, place, etc.)  
- They are very useful in NLP tasks like NER  

### **3. Which preprocessing step may harm entity extraction?**
Poor tokenization can create problems  

**Examples:**  
- "Dr. Smith" → ["Dr.", "Smith"] (connection between title and name is weak)  
- "HR_department" → ["HR_department"] (not split into clear words)  
- "10:30 AM" → ["10:30", "AM"] (time is split and may confuse model)  

**Problems caused:**  
- Breaks connection between related words  
- Does not split words properly  
- Reduces accuracy in entity recognition  

</div>

#----------------------------------------------------------------------------------------------------------------------------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **4. Stopword Removal**

### **What is this technique?**
Stopwords are common words like "is", "the", "and", "to", "in".  
These words do not add much meaning to the sentence.

**Example:**  
"I am going to the market" → ["going", "market"]

### **Why is it used?**
- Removes unnecessary words  
- Reduces data size  
- Helps model focus on important words  

</div>

In [11]:
raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!",
    "Hi!!! How are you doing today??? 😊",
    "Complete the project ASAP!!! Deadline is 25/12/2025.",
    "Let's catch up at Starbucks, Chennai.",
    "Call mom @ 7pm sharp.",
    "Meeting rescheduled from 5pm to 6:30pm.",
    "Send the EMAIL to HR_department immediately.",
    "Book flight tickets from Chennai to Delhi.",
    "Wake me up at 6 in the morning.",
    "Don't forget to bring the documents!!!",
    "Can we meet next Friday at 4pm?"
]

stopwords = ["i", "am", "a", "the", "is", "are", "to", "at", "me", "you", "we", "can", "be", "in", "on", "up", "how"]

def remove_stopwords(text):
    words = text.lower().split()
    filtered_words = [word for word in words if word not in stopwords]
    return filtered_words

filtered_texts = [remove_stopwords(text) for text in raw_texts]

In [12]:
for before, after in zip(raw_texts, filtered_texts):
    print("BEFORE :", before)
    print("AFTER  :", after)
    print("-" * 50)

BEFORE : I HAVE a meeting @ 3pm!!!
AFTER  : ['have', 'meeting', '@', '3pm!!!']
--------------------------------------------------
BEFORE : Hey John, can u pls call me tomorrow??
AFTER  : ['hey', 'john,', 'u', 'pls', 'call', 'tomorrow??']
--------------------------------------------------
BEFORE : Reminder: Pay electricity bill before 10th Sept.
AFTER  : ['reminder:', 'pay', 'electricity', 'bill', 'before', '10th', 'sept.']
--------------------------------------------------
BEFORE : Schedule meeting with Dr. Smith at 10:30 AM.
AFTER  : ['schedule', 'meeting', 'with', 'dr.', 'smith', '10:30', 'am.']
--------------------------------------------------
BEFORE : Buy 2kg apples, 3 bananas & some milk!!
AFTER  : ['buy', '2kg', 'apples,', '3', 'bananas', '&', 'some', 'milk!!']
--------------------------------------------------
BEFORE : Hi!!! How are you doing today??? 😊
AFTER  : ['hi!!!', 'doing', 'today???', '😊']
--------------------------------------------------
BEFORE : Complete the project 

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

### **What changed?**
- Common words like "I", "a", "the", "to" are removed  
- Only important words are kept  

**Example:**  
"I HAVE a meeting @ 3pm!!!" → ["have", "meeting", "@", "3pm!!!"]

### **Did the meaning change?**
- Slightly reduced, but main meaning is still clear  
- Important words are still present  

### **Is this good or bad for NLP?**

Good:  
- Reduces noise in data  
- Focuses on important words  
- Improves model performance  

Limitation:  
- Removes some useful words  
- Sentence may lose natural structure  

</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No  
- "3pm", "7pm", "10:30", "25/12/2025" give important time and date information  

### **2. Should we remove names like "John"?**
No  
- "John", "Smith", "Chennai", "Delhi" are important entities  
- Useful for tasks like NER  

### **3. Which preprocessing step may harm entity extraction?**
Stopword removal can also affect entity meaning  

**Examples:**  
- "Call me tomorrow" → ["call", "tomorrow"] (lost "me")  
- "Meet me at Chennai" → ["meet", "chennai"] (lost relation word "at")  
- "Send the EMAIL to HR_department" → ["send", "email", "hr_department"]  

**Problems caused:**  
- Removes connecting words  
- Sentence meaning may become unclear  
- May affect relationship between entities  

</div>

#----------------------------------------------------------------------------------------------------------------------------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **5. Stemming**

### **What is this technique?**
Stemming means reducing a word to its root (base) form.

**Example:**  
"playing", "played", "plays" → "play"

### **Why is it used?**
- Reduces different forms of a word  
- Helps treat similar words as same  
- Reduces vocabulary size  

</div>

In [13]:
from nltk.stem import PorterStemmer

stemmer = PorterStemmer()

raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!",
    "Hi!!! How are you doing today??? 😊",
    "Complete the project ASAP!!! Deadline is 25/12/2025.",
    "Let's catch up at Starbucks, Chennai.",
    "Call mom @ 7pm sharp.",
    "Meeting rescheduled from 5pm to 6:30pm.",
    "Send the EMAIL to HR_department immediately.",
    "Book flight tickets from Chennai to Delhi.",
    "Wake me up at 6 in the morning.",
    "Don't forget to bring the documents!!!",
    "Can we meet next Friday at 4pm?"
]

def apply_stemming(text):
    words = text.lower().split()
    stemmed_words = [stemmer.stem(word) for word in words]
    return stemmed_words

stemmed_texts = [apply_stemming(text) for text in raw_texts]

In [14]:
for before, after in zip(raw_texts, stemmed_texts):
    print("BEFORE :", before)
    print("AFTER  :", after)
    print("-" * 50)

BEFORE : I HAVE a meeting @ 3pm!!!
AFTER  : ['i', 'have', 'a', 'meet', '@', '3pm!!!']
--------------------------------------------------
BEFORE : Hey John, can u pls call me tomorrow??
AFTER  : ['hey', 'john,', 'can', 'u', 'pl', 'call', 'me', 'tomorrow??']
--------------------------------------------------
BEFORE : Reminder: Pay electricity bill before 10th Sept.
AFTER  : ['reminder:', 'pay', 'electr', 'bill', 'befor', '10th', 'sept.']
--------------------------------------------------
BEFORE : Schedule meeting with Dr. Smith at 10:30 AM.
AFTER  : ['schedul', 'meet', 'with', 'dr.', 'smith', 'at', '10:30', 'am.']
--------------------------------------------------
BEFORE : Buy 2kg apples, 3 bananas & some milk!!
AFTER  : ['buy', '2kg', 'apples,', '3', 'banana', '&', 'some', 'milk!!']
--------------------------------------------------
BEFORE : Hi!!! How are you doing today??? 😊
AFTER  : ['hi!!!', 'how', 'are', 'you', 'do', 'today???', '😊']
-------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

### **What changed?**
- Words are reduced to their root form  
- Example:  
  "meeting" → "meet"  
  "doing" → "do"  
  "tickets" → "ticket"  

### **Did the meaning change?**
- Slightly, yes  
- Some words may not look natural after stemming  

### **Is this good or bad for NLP?**

Good:  
- Reduces similar words into one form  
- Helps model learn better  
- Improves performance in many tasks  

Limitation:  
- Words may become incorrect (e.g., "meeting" → "meet")  
- Output is not always meaningful English  

</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No  
- "3pm", "7pm", "10:30", "25/12/2025" are important  
- They give time and date information  

### **2. Should we remove names like "John"?**
No  
- "John", "Smith", "Chennai", "Delhi" are important entities  
- Useful in NLP tasks like NER  

### **3. Which preprocessing step may harm entity extraction?**
Stemming can harm entity meaning  

**Examples:**  
- "meeting" → "meet" (changes word form)  
- "Chennai" → "chennai" (may lose proper noun format)  
- "documents" → "document" (changes structure)  

**Problems caused:**  
- Words lose original form  
- May affect entity recognition  
- Output may not be natural English  

</div>

#----------------------------------------------------------------------------------------------------------------------------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **6. Lemmatization**

### **What is this technique?**
Lemmatization means converting a word into its correct base (dictionary) form.

**Example:**  
"running", "ran", "runs" → "run"  

### **Why is it used?**
- Converts words into meaningful base form  
- Keeps correct English words  
- Better than stemming for accuracy  

</div>

In [15]:
import nltk
from nltk.stem import WordNetLemmatizer

nltk.download('wordnet')

lemmatizer = WordNetLemmatizer()

raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!",
    "Hi!!! How are you doing today??? 😊",
    "Complete the project ASAP!!! Deadline is 25/12/2025.",
    "Let's catch up at Starbucks, Chennai.",
    "Call mom @ 7pm sharp.",
    "Meeting rescheduled from 5pm to 6:30pm.",
    "Send the EMAIL to HR_department immediately.",
    "Book flight tickets from Chennai to Delhi.",
    "Wake me up at 6 in the morning.",
    "Don't forget to bring the documents!!!",
    "Can we meet next Friday at 4pm?"
]

def apply_lemmatization(text):
    words = text.lower().split()
    lemmatized_words = [lemmatizer.lemmatize(word) for word in words]
    return lemmatized_words

lemmatized_texts = [apply_lemmatization(text) for text in raw_texts]

[nltk_data] Downloading package wordnet to /root/nltk_data...


In [16]:
for before, after in zip(raw_texts, lemmatized_texts):
    print("BEFORE :", before)
    print("AFTER  :", after)
    print("-" * 50)

BEFORE : I HAVE a meeting @ 3pm!!!
AFTER  : ['i', 'have', 'a', 'meeting', '@', '3pm!!!']
--------------------------------------------------
BEFORE : Hey John, can u pls call me tomorrow??
AFTER  : ['hey', 'john,', 'can', 'u', 'pls', 'call', 'me', 'tomorrow??']
--------------------------------------------------
BEFORE : Reminder: Pay electricity bill before 10th Sept.
AFTER  : ['reminder:', 'pay', 'electricity', 'bill', 'before', '10th', 'sept.']
--------------------------------------------------
BEFORE : Schedule meeting with Dr. Smith at 10:30 AM.
AFTER  : ['schedule', 'meeting', 'with', 'dr.', 'smith', 'at', '10:30', 'am.']
--------------------------------------------------
BEFORE : Buy 2kg apples, 3 bananas & some milk!!
AFTER  : ['buy', '2kg', 'apples,', '3', 'banana', '&', 'some', 'milk!!']
--------------------------------------------------
BEFORE : Hi!!! How are you doing today??? 😊
AFTER  : ['hi!!!', 'how', 'are', 'you', 'doing', 'today???', '😊']
--------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

### **What changed?**
- Words are converted into their correct base form  
- Example:  
  "meeting" → "meeting"  
  "doing" → "doing"  
  "tickets" → "ticket"  

### **Did the meaning change?**
- No major change  
- Words remain meaningful and correct  

### **Is this good or bad for NLP?**

Good:  
- Keeps correct English words  
- Better accuracy than stemming  
- Useful for real-world NLP applications  

Limitation:  
- Slightly slower than stemming  
- Needs proper context (POS tagging) for best results  

</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No  
- "3pm", "7pm", "10:30", "25/12/2025" are important  
- They give time and date information  

### **2. Should we remove names like "John"?**
No  
- "John", "Smith", "Chennai", "Delhi" are important entities  
- Useful in NLP tasks like NER  

### **3. Which preprocessing step may harm entity extraction?**
Lemmatization (if used incorrectly) can affect entities  

**Examples:**  
- "documents" → "document" (changes original form)  
- "tickets" → "ticket" (number information lost)  
- "meeting" → "meeting" (sometimes no change)  

**Problems caused:**  
- May change word form  
- Can lose plural/singular meaning  
- Needs context (POS) for correct output  

</div>

#----------------------------------------------------------------------------------------------------------------------------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **7. Removing Numbers**

### **What is this technique?**
Removing numbers means deleting digits like 0–9 from the text.

**Example:**  
"I have 2 apples" → "I have apples"

### **Why is it used?**
- Removes unnecessary numeric data  
- Simplifies text  
- Useful when numbers are not important  

</div>

In [17]:
import re

raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!",
    "Hi!!! How are you doing today??? 😊",
    "Complete the project ASAP!!! Deadline is 25/12/2025.",
    "Let's catch up at Starbucks, Chennai.",
    "Call mom @ 7pm sharp.",
    "Meeting rescheduled from 5pm to 6:30pm.",
    "Send the EMAIL to HR_department immediately.",
    "Book flight tickets from Chennai to Delhi.",
    "Wake me up at 6 in the morning.",
    "Don't forget to bring the documents!!!",
    "Can we meet next Friday at 4pm?"
]

def remove_numbers(text):
    return re.sub(r'\d+', '', text)

no_number_texts = [remove_numbers(text) for text in raw_texts]

In [18]:
for before, after in zip(raw_texts, no_number_texts):
    print("BEFORE :", before)
    print("AFTER  :", after)
    print("-" * 50)

BEFORE : I HAVE a meeting @ 3pm!!!
AFTER  : I HAVE a meeting @ pm!!!
--------------------------------------------------
BEFORE : Hey John, can u pls call me tomorrow??
AFTER  : Hey John, can u pls call me tomorrow??
--------------------------------------------------
BEFORE : Reminder: Pay electricity bill before 10th Sept.
AFTER  : Reminder: Pay electricity bill before th Sept.
--------------------------------------------------
BEFORE : Schedule meeting with Dr. Smith at 10:30 AM.
AFTER  : Schedule meeting with Dr. Smith at : AM.
--------------------------------------------------
BEFORE : Buy 2kg apples, 3 bananas & some milk!!
AFTER  : Buy kg apples,  bananas & some milk!!
--------------------------------------------------
BEFORE : Hi!!! How are you doing today??? 😊
AFTER  : Hi!!! How are you doing today??? 😊
--------------------------------------------------
BEFORE : Complete the project ASAP!!! Deadline is 25/12/2025.
AFTER  : Complete the project ASAP!!! Deadline is //.
-----------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

### **What changed?**
- All numbers are removed from the text  
- Example:  
  "3pm" → "pm"  
  "10:30" → ":"  
  "25/12/2025" → "//"  

### **Did the meaning change?**
- Yes, in many cases  
- Important information like time and date is lost  

### **Is this good or bad for NLP?**

Good:  
- Useful when numbers are not needed  
- Reduces noise in some datasets  

Limitation:  
- Removes important information  
- Can break meaning of sentences  

</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

### **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No  
- "3pm", "7pm", "10:30", "25/12/2025" contain important information  
- Removing numbers removes time and date meaning  

### **2. Should we remove names like "John"?**
No  
- "John", "Smith", "Chennai", "Delhi" are important entities  
- Needed for NLP tasks like NER  

### **3. Which preprocessing step may harm entity extraction?**
Removing numbers can harm entity extraction  

**Examples:**  
- "3pm" → "pm" (time lost)  
- "10:30 AM" → ": AM" (invalid format)  
- "25/12/2025" → "//" (date completely lost)  

**Problems caused:**  
- Removes important numeric information  
- Breaks time and date formats  
- Reduces accuracy in real-world applications  

</div>

#----------------------------------------------------------------------------------------------------------------------------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **8. Removing Special Characters**

### **What is this technique?**
Removing special characters means deleting symbols like @, #, &, %, emojis, etc. from text.

**Example:**  
"Hello @user!!! 😊" → "Hello user"

### **Why is it used?**
- Removes unwanted symbols  
- Cleans the text  
- Makes data easier for models to understand  

</div>

In [19]:
import re

raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!",
    "Hi!!! How are you doing today??? 😊",
    "Complete the project ASAP!!! Deadline is 25/12/2025.",
    "Let's catch up at Starbucks, Chennai.",
    "Call mom @ 7pm sharp.",
    "Meeting rescheduled from 5pm to 6:30pm.",
    "Send the EMAIL to HR_department immediately.",
    "Book flight tickets from Chennai to Delhi.",
    "Wake me up at 6 in the morning.",
    "Don't forget to bring the documents!!!",
    "Can we meet next Friday at 4pm?"
]

def remove_special_characters(text):
    return re.sub(r'[^a-zA-Z0-9\s]', '', text)

cleaned_texts = [remove_special_characters(text) for text in raw_texts]

In [20]:
for before, after in zip(raw_texts, cleaned_texts):
    print("BEFORE :", before)
    print("AFTER  :", after)
    print("-" * 50)

BEFORE : I HAVE a meeting @ 3pm!!!
AFTER  : I HAVE a meeting  3pm
--------------------------------------------------
BEFORE : Hey John, can u pls call me tomorrow??
AFTER  : Hey John can u pls call me tomorrow
--------------------------------------------------
BEFORE : Reminder: Pay electricity bill before 10th Sept.
AFTER  : Reminder Pay electricity bill before 10th Sept
--------------------------------------------------
BEFORE : Schedule meeting with Dr. Smith at 10:30 AM.
AFTER  : Schedule meeting with Dr Smith at 1030 AM
--------------------------------------------------
BEFORE : Buy 2kg apples, 3 bananas & some milk!!
AFTER  : Buy 2kg apples 3 bananas  some milk
--------------------------------------------------
BEFORE : Hi!!! How are you doing today??? 😊
AFTER  : Hi How are you doing today 
--------------------------------------------------
BEFORE : Complete the project ASAP!!! Deadline is 25/12/2025.
AFTER  : Complete the project ASAP Deadline is 25122025
-----------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

### **What changed?**
- Special characters like "@", "!", "😊", "&" are removed  
- Example:  
  "@ 3pm!!!" → " 3pm"  
  "Hi!!! 😊" → "Hi"  

### **Did the meaning change?**
- Slightly, yes  
- Some symbols may carry meaning (like "@", "&")  

### **Is this good or bad for NLP?**

Good:  
- Cleans noisy data  
- Makes text simple  
- Useful for many NLP tasks  

Limitation:  
- Removes useful symbols  
- May affect meaning in some cases  

</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No  
- "3pm", "7pm", "10:30", "25/12/2025" contain important information  

### **2. Should we remove names like "John"?**
No  
- "John", "Smith", "Chennai", "Delhi" are important entities  

### **3. Which preprocessing step may harm entity extraction?**
Removing special characters can affect entities  

**Examples:**  
- "Dr. Smith" → "Dr Smith" (dot removed)  
- "HR_department" → "HRdepartment" (structure lost)  
- "10:30 AM" → "1030 AM" (time format changed)  

**Problems caused:**  
- Removes useful structure  
- Changes format of important data  
- Reduces accuracy in entity recognition  

</div>

#----------------------------------------------------------------------------------------------------------------------------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **9. Removing Extra Whitespaces**

### **What is this technique?**
Removing extra whitespaces means cleaning unnecessary spaces in text.

**Example:**  
"I   have   a   meeting" → "I have a meeting"


"Call   mom   at   7pm" → "Call mom at 7pm"

### **Why is it used?**
- Cleans messy text  
- Ensures uniform spacing  
- Improves readability  

</div>

In [21]:
import re

raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!",
    "Hi!!! How are you doing today??? 😊",
    "Complete the project ASAP!!! Deadline is 25/12/2025.",
    "Let's catch up at Starbucks, Chennai.",
    "Call mom @ 7pm sharp.",
    "Meeting rescheduled from 5pm to 6:30pm.",
    "Send the EMAIL to HR_department immediately.",
    "Book flight tickets from Chennai to Delhi.",
    "Wake me up at 6 in the morning.",
    "Don't forget to bring the documents!!!",
    "Can we meet next Friday at 4pm?"
]

def remove_extra_whitespace(text):
    return re.sub(r'\s+', ' ', text).strip()

clean_whitespace_texts = [remove_extra_whitespace(text) for text in raw_texts]

In [22]:
for before, after in zip(raw_texts, clean_whitespace_texts):
    print("BEFORE :", before)
    print("AFTER  :", after)
    print("-" * 50)

BEFORE : I HAVE a meeting @ 3pm!!!
AFTER  : I HAVE a meeting @ 3pm!!!
--------------------------------------------------
BEFORE : Hey John, can u pls call me tomorrow??
AFTER  : Hey John, can u pls call me tomorrow??
--------------------------------------------------
BEFORE : Reminder: Pay electricity bill before 10th Sept.
AFTER  : Reminder: Pay electricity bill before 10th Sept.
--------------------------------------------------
BEFORE : Schedule meeting with Dr. Smith at 10:30 AM.
AFTER  : Schedule meeting with Dr. Smith at 10:30 AM.
--------------------------------------------------
BEFORE : Buy 2kg apples, 3 bananas & some milk!!
AFTER  : Buy 2kg apples, 3 bananas & some milk!!
--------------------------------------------------
BEFORE : Hi!!! How are you doing today??? 😊
AFTER  : Hi!!! How are you doing today??? 😊
--------------------------------------------------
BEFORE : Complete the project ASAP!!! Deadline is 25/12/2025.
AFTER  : Complete the project ASAP!!! Deadline is 25/12/

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

### **What changed?**
- Multiple spaces are reduced to a single space  
- Leading and trailing spaces are removed  

**Example:**  
"I   have   a meeting" → "I have a meeting"

### **Did the meaning change?**
- No  
- Only formatting is improved  

### **Is this good or bad for NLP?**

Good:  
- Improves text quality  
- Makes data consistent  
- Safe preprocessing step  

Limitation:  
- No major limitation  
- Does not affect meaning  

</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No  
- "3pm", "7pm", "10:30", "25/12/2025" contain important information  

### **2. Should we remove names like "John"?**
No  
- "John", "Smith", "Chennai", "Delhi" are important entities  

### **3. Which preprocessing step may harm entity extraction?**
Removing extra whitespaces usually does NOT harm entity extraction  

**Examples:**  
- "Dr.   Smith" → "Dr. Smith" (corrected spacing)  
- "Chennai   to   Delhi" → "Chennai to Delhi"  
- "Meeting   at   3pm" → "Meeting at 3pm"  

**Benefits:**  
- Improves text clarity  
- Keeps meaning unchanged  
- Helps better processing  

</div>

#----------------------------------------------------------------------------------------------------------------------------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **10. N-grams (Unigram, Bigram, Trigram)**

### **What is this technique?**
N-grams means grouping words together based on sequence.

- Unigram → single word  
- Bigram → group of 2 words  
- Trigram → group of 3 words  

**Example:**  
"I have a meeting"  

Unigram → ["I", "have", "a", "meeting"]  
Bigram → ["I have", "have a", "a meeting"]  
Trigram → ["I have a", "have a meeting"]

### **Why is it used?**
- Captures word relationships  
- Helps understand context  
- Improves model performance  

</div>

In [23]:
from nltk.util import ngrams

raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!",
    "Hi!!! How are you doing today??? 😊",
    "Complete the project ASAP!!! Deadline is 25/12/2025.",
    "Let's catch up at Starbucks, Chennai.",
    "Call mom @ 7pm sharp.",
    "Meeting rescheduled from 5pm to 6:30pm.",
    "Send the EMAIL to HR_department immediately.",
    "Book flight tickets from Chennai to Delhi.",
    "Wake me up at 6 in the morning.",
    "Don't forget to bring the documents!!!",
    "Can we meet next Friday at 4pm?"
]

def generate_ngrams(text, n):
    words = text.lower().split()
    return list(ngrams(words, n))

sample_text = raw_texts[0]

unigram = generate_ngrams(sample_text, 1)
bigram = generate_ngrams(sample_text, 2)
trigram = generate_ngrams(sample_text, 3)

print("Sentence:", sample_text)
print("Unigram:", unigram)
print("Bigram:", bigram)
print("Trigram:", trigram)

Sentence: I HAVE a meeting @ 3pm!!!
Unigram: [('i',), ('have',), ('a',), ('meeting',), ('@',), ('3pm!!!',)]
Bigram: [('i', 'have'), ('have', 'a'), ('a', 'meeting'), ('meeting', '@'), ('@', '3pm!!!')]
Trigram: [('i', 'have', 'a'), ('have', 'a', 'meeting'), ('a', 'meeting', '@'), ('meeting', '@', '3pm!!!')]


<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

### **What changed?**
- Words are grouped into sequences  
- Example:  
  "I HAVE a meeting"  
  → Unigram: single words  
  → Bigram: pairs of words  
  → Trigram: groups of 3 words  

### **Did the meaning change?**
- No, meaning is not lost  
- Instead, more context is captured  

### **Is this good or bad for NLP?**

Good:  
- Captures word relationships  
- Improves context understanding  
- Useful in language models  

Limitation:  
- Increases data size  
- Higher n-grams need more memory  
- May create sparse data  

</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No  
- "3pm", "7pm", "10:30", "25/12/2025" contain important information  
- Useful in forming meaningful n-grams  

### **2. Should we remove names like "John"?**
No  
- "John", "Smith", "Chennai", "Delhi" are important  
- Helps create meaningful phrases like "John call", "Chennai to Delhi"  

### **3. Which preprocessing step may harm entity extraction?**
N-grams combined with bad preprocessing can harm entities  

**Examples:**  
- "Dr Smith" → Bigram: ("dr", "smith") (may lose title importance)  
- "Chennai to Delhi" → Bigram: ("chennai", "to"), ("to", "delhi")  
- "HR_department" → treated as one token (structure issue)  

**Problems caused:**  
- Wrong grouping of words  
- Loss of entity meaning  
- Increased complexity  

</div>

#----------------------------------------------------------------------------------------------------------------------------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **11. Bag of Words (BoW)**

### **What is this technique?**
Bag of Words converts text into numbers based on word frequency.

- It ignores grammar and word order  
- It only counts how many times words appear  

**Example:**  
"I have a meeting"  
"I have a call"  

Vocabulary → ["I", "have", "a", "meeting", "call"]  

Vector Representation:  
"I have a meeting" → [1, 1, 1, 1, 0]  
"I have a call" → [1, 1, 1, 0, 1]

### **Why is it used?**
- Converts text into numerical format  
- Easy to implement  
- Works well for basic NLP models  

</div>

In [26]:
from sklearn.feature_extraction.text import CountVectorizer

raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!"
]

vectorizer = CountVectorizer()

X = vectorizer.fit_transform(raw_texts)

bow_array = X.toarray()

features = vectorizer.get_feature_names_out()

print("Vocabulary:", features)
print("\nBoW Matrix:\n", bow_array)

Vocabulary: ['10' '10th' '2kg' '30' '3pm' 'am' 'apples' 'at' 'bananas' 'before' 'bill'
 'buy' 'call' 'can' 'dr' 'electricity' 'have' 'hey' 'john' 'me' 'meeting'
 'milk' 'pay' 'pls' 'reminder' 'schedule' 'sept' 'smith' 'some' 'tomorrow'
 'with']

BoW Matrix:
 [[0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 1 1 1 0 0 0 1 0 0 0 0 0 1 0]
 [0 1 0 0 0 0 0 0 0 1 1 0 0 0 0 1 0 0 0 0 0 0 1 0 1 0 1 0 0 0 0]
 [1 0 0 1 0 1 0 1 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 1 0 1 0 0 1]
 [0 0 1 0 0 0 1 0 1 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0]]


In [27]:
print("Sentence 1:", raw_texts[0])
print("Vector:", bow_array[0])

Sentence 1: I HAVE a meeting @ 3pm!!!
Vector: [0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0]


<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

### **What changed?**
- Text is converted into numbers  
- Each word becomes a column (feature)  
- Values show word count  

### **Did the meaning change?**
- Partially yes  
- Word order is lost  
- Only frequency is kept  

### **Is this good or bad for NLP?**

Good:  
- Simple and fast  
- Works well for basic models  
- Easy to understand  

Limitation:  
- Ignores word order  
- Cannot capture context  
- Large vocabulary increases size  

</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No  
- "3pm", "10:30", "25/12/2025" are useful  
- They become features in BoW  

### **2. Should we remove names like "John"?**
No  
- Names like "John", "Smith", "Chennai" are important  
- They help in identifying entities  

### **3. Which preprocessing step may harm BoW?**
Removing important words or poor cleaning  

**Examples:**  
- Removing "not" → changes meaning ("not good" → "good")  
- Removing "John" → loss of entity  
- Removing numbers → loss of time/date info  

**Problems caused:**  
- Loss of important meaning  
- Wrong predictions  
- Reduced model accuracy  

</div>

#----------------------------------------------------------------------------------------------------------------------------------------------------------------------

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **12. TF-IDF (Term Frequency - Inverse Document Frequency)**

### **What is this technique?**
TF-IDF gives importance to words based on how useful they are.

- TF (Term Frequency) → how often a word appears in a sentence  
- IDF (Inverse Document Frequency) → how rare the word is across all sentences  

Common words → low score  
Rare important words → high score  

**Example:**  
"I have a meeting"  
"I have a call"  

- "have", "a" → common → low importance  
- "meeting", "call" → rare → high importance  

### **Why is it used?**
- Highlights important words  
- Reduces impact of common words  
- Better than Bag of Words  

</div>

In [28]:
from sklearn.feature_extraction.text import TfidfVectorizer

raw_texts = [
    "I HAVE a meeting @ 3pm!!!",
    "Hey John, can u pls call me tomorrow??",
    "Reminder: Pay electricity bill before 10th Sept.",
    "Schedule meeting with Dr. Smith at 10:30 AM.",
    "Buy 2kg apples, 3 bananas & some milk!!"
]

vectorizer = TfidfVectorizer()

X = vectorizer.fit_transform(raw_texts)

tfidf_array = X.toarray()

features = vectorizer.get_feature_names_out()

print("Vocabulary:", features)
print("\nTF-IDF Matrix:\n", tfidf_array)

Vocabulary: ['10' '10th' '2kg' '30' '3pm' 'am' 'apples' 'at' 'bananas' 'before' 'bill'
 'buy' 'call' 'can' 'dr' 'electricity' 'have' 'hey' 'john' 'me' 'meeting'
 'milk' 'pay' 'pls' 'reminder' 'schedule' 'sept' 'smith' 'some' 'tomorrow'
 'with']

TF-IDF Matrix:
 [[0.         0.         0.         0.         0.61418897 0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.61418897 0.
  0.         0.         0.49552379 0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.        ]
 [0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.         0.         0.
  0.37796447 0.37796447 0.         0.         0.         0.37796447
  0.37796447 0.37796447 0.         0.         0.         0.37796447
  0.         0.         0.         0.         0.         0.37796447
  0.        ]
 [0.         0.37796447 0.         0.         0.         0.
  0.         0.         0.

In [29]:
print("Sentence 1:", raw_texts[0])
print("TF-IDF Vector:", tfidf_array[0])

Sentence 1: I HAVE a meeting @ 3pm!!!
TF-IDF Vector: [0.         0.         0.         0.         0.61418897 0.
 0.         0.         0.         0.         0.         0.
 0.         0.         0.         0.         0.61418897 0.
 0.         0.         0.49552379 0.         0.         0.
 0.         0.         0.         0.         0.         0.
 0.        ]


<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

### **What changed?**
- Text is converted into weighted numbers  
- Important words get higher values  
- Common words get lower values  

### **Did the meaning change?**
- No, meaning is preserved better than BoW  
- Important words are highlighted  

### **Is this good or bad for NLP?**

Good:  
- Better than Bag of Words  
- Reduces noise from common words  
- Improves model accuracy  

Limitation:  
- Still ignores word order  
- Cannot fully capture context  
- Slightly more complex than BoW  

</div>

<div style="font-family: 'Gill Sans', 'Segoe UI', Arial, sans-serif; font-size:16px; line-height:1.6;">

# **Critical Thinking**

### **1. Should we remove numbers like "3pm"?**
No  
- "3pm", "10:30", "25/12/2025" may carry important meaning  
- TF-IDF can assign importance if they are rare  

### **2. Should we remove names like "John"?**
No  
- Names like "John", "Smith", "Chennai" are important  
- They may get high TF-IDF scores  

### **3. Which preprocessing step may harm TF-IDF?**
Removing important or rare words  

**Examples:**  
- Removing "meeting" → lose key information  
- Removing "John" → lose entity  
- Removing numbers → lose time/date info  

**Problems caused:**  
- Important words lost  
- Lower model accuracy  
- Poor feature representation  

</div>

##Thank You !